analyzers/src/SonarAnalyzer.CSharp/Rules/MessageTemplatesShouldBeCorrect.cs

```diff
@@ -0,0 +1,160 @@
+            {
+                return new($"should not contain empty placeholder", placeholder.Start - 1, 2); // highlight both brackets '{}', since there is no content
+            }
+
+            var parts = Split(placeholder.Name);
+            if (!PlaceholderNameRegex.SafeIsMatch(parts.Name))
+            {
+                return new($"placeholder '{parts.Name}' should only contain chars, numbers, and underscore", placeholder);
+            }
+            else if (parts.Alignment is not null && !PlaceholderAlignmentRegex.SafeIsMatch(parts.Alignment))
+            {
+                return new($"placeholder '{parts.Name}' should have numeric alignment instead of '{parts.Alignment}'", placeholder);
+            }
+            else if (parts.Format == string.Empty)
+            {
+                return new($"placeholder '{parts.Name}' should not have empty format", placeholder);
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        // pattern is: name[,alignment][:format]
+        private static Parts Split(string placeholder)
+        {
+            var start = placeholder[0] is '@' or '$' ? 1 : 0; // skip prefix
+
+            string name;
+            string alignment = null;
+            string format = null;
+
+            var formatIndex = placeholder.IndexOf(':');
+            var alignmentIndex = placeholder.IndexOf(',');
+            if (formatIndex >= 0 && alignmentIndex > formatIndex)
+            {
+                // example {name:format,alignment}
+                //               ^^^^^^^^^^^^^^^^ all of this is format, need to reset alignment
+                alignmentIndex = -1;
+            }
+
+            if (NotFound(alignmentIndex))
+            {
+                if (NotFound(formatIndex))
+                {
+                    name = placeholder.Substring(start);
+                }
+                else
+                {
+                    name = placeholder.Substring(start, formatIndex - start);
+                    format = IsEmpty(formatIndex) ? string.Empty : placeholder.Substring(formatIndex + 1);
+                }
+            }
+            else
+            {
+                if (NotFound(formatIndex))
+                {
+                    name = placeholder.Substring(start, alignmentIndex - start);
+                    alignment = IsEmpty(alignmentIndex) ? string.Empty : placeholder.Substring(alignmentIndex + 1);
+                }
+                else
+                {
+                    name = placeholder.Substring(start, alignmentIndex - start);
+                    alignment = placeholder.Substring(alignmentIndex + 1, formatIndex - alignmentIndex - 1);
+                    format = placeholder.Substring(formatIndex + 1);
+                }
+            }
+
+            return new(name, alignment, format);
+
+            bool NotFound(int index) => index == -1;
+            bool IsEmpty(int index) => index == placeholder.Length - 1;
```

Comment:
The way I see it is that we have a segment that we have to split into three parts. We can start by working our way from the end since it's easier:
```
            if (formatIndex == -1)
            {
                formatIndex = placeholder.Length;
            }
            else
            {
                format = placeholder.Substring(formatIndex + 1);
            }
            if (alignmentIndex == -1)
            {
                alignmentIndex = formatIndex;
            }
            else
            {
                alignment = placeholder.Substring(alignmentIndex + 1, formatIndex - alignmentIndex - 1);
            }
            name = placeholder.Substring(start, alignmentIndex - start);
```
This should also take care of the code smell raised for this function.

Reply:
Excellent idea :O
So much cleaner than my implementation. 
Thanks a lot!